pip/pep425tags.py

```diff
@@ -187,27 +188,97 @@ def get_supported(versions=None, noarch=False):
+        # Exact match for ABI and platform
+        for version in versions[1:]:
+            # Explicit Python
+            supported.append(('%s%s' % (impl, version), abis[0], arches[0]))
+            # Generic Python
+            supported.append(('py%s' % (version,), abis[0], arches[0]))
+        # Exact match for Python
+        for abi in abis[1:]:
+            for arch in arches[1:]:
+                # Explicit Python
                 supported.append(('%s%s' % (impl, versions[0]), abi, arch))
-
-        # Has binaries, does not use the Python API:
-        supported.append(('py%s' % (versions[0][0]), 'none', arch))
-
-    # No abi / arch, but requires our implementation:
-    for i, version in enumerate(versions):
-        supported.append(('%s%s' % (impl, version), 'none', 'any'))
-        if i == 0:
-            # Tagged specifically as being cross-version compatible
-            # (with just the major version specified)
-            supported.append(('%s%s' % (impl, versions[0][0]), 'none', 'any'))
-
-    # No abi / arch, generic Python
-    for i, version in enumerate(versions):
-        supported.append(('py%s' % (version,), 'none', 'any'))
-        if i == 0:
-            supported.append(('py%s' % (version[0]), 'none', 'any'))
+                supported.append(('%s%s' % (impl, versions[0][0]), abi, arch))
+                # Generic Python
+                supported.append(('py%s' % (versions[0]), abi, arch))
+                supported.append(('py%s' % (versions[0][0]), abi, arch))
+        # Exact match for ABI
+        for version in versions[1:]:
+            for arch in arches[1:]:
+                # Explicit Python
+                supported.append(('%s%s' % (impl, version), abis[0], arch))
+                # Generic Python
+                supported.append(('py%s' % (version,), abis[0], arch))
+        # Exact match for platform
+        for version in versions[1:]:
+            for abi in abis[1:]:
+                # Explicit Python
+                supported.append(('%s%s' % (impl, version), abi, arches[0]))
+                # Generic Python
+                supported.append(('py%s' % (version,), abi, arches[0]))
+        # No exact match
+        for version in versions[1:]:
+            for abi in abis[1:]:
+                for arch in arches[1:]:
+                    # Explicit Python
+                    supported.append(('%s%s' % (impl, version), abi, arch))
+                    # Generic Python
+                    supported.append(('py%s' % (version,), abi, arch))
+    else:
+        # Exact match for Python and ABI
+        # Explicit Python
+        supported.append(('%s%s' % (impl, versions[0]), abis[0], 'any'))
+        supported.append(('%s%s' % (impl, versions[0][0]), abis[0], 'any'))
+        # Generic Python
+        supported.append(('py%s' % (versions[0]), abis[0], 'any'))
+        supported.append(('py%s' % (versions[0][0]), abis[0], 'any'))
```

Comment:
Good point, I've removed them and removed the `[1:]` from `abis` in the next loop.